File: src/main/java/io/bindernews/thegrackle/power/TempThornsPower.java
package io.bindernews.thegrackle.power;

import com.megacrit.cardcrawl.actions.AbstractGameAction;
import com.megacrit.cardcrawl.actions.common.DamageAction;
import com.megacrit.cardcrawl.actions.common.RemoveSpecificPowerAction;
import com.megacrit.cardcrawl.cards.DamageInfo;
import com.megacrit.cardcrawl.cards.DamageInfo.DamageType;
import com.megacrit.cardcrawl.core.AbstractCreature;
import io.bindernews.thegrackle.GrackleMod;

public class TempThornsPower extends BasePower {
    public static final String POWER_ID = GrackleMod.makeId(TempThornsPower.class);

    public TempThornsPower(AbstractCreature owner, int amount) {
        super(POWER_ID);
        setOwnerAmount(owner, amount);
        loadRegion("thorns");
        isTurnBased = true;
        updateDescription();
    }

    public int onAttacked(DamageInfo info, int damageAmount) {
        if (info.type != DamageType.THORNS
                && info.type != DamageType.HP_LOSS
                && info.owner != null
                && info.owner != this.owner) {
            this.flash();
            this.addToTop(new DamageAction(info.owner, new DamageInfo(this.owner, this.amount, DamageType.THORNS), AbstractGameAction.AttackEffect.SLASH_HORIZONTAL, true));
        }

        return damageAmount;
    }

    public void updateDescription() {
        description = formatDesc(0, amount);
    }

    @Override
    public void atStartOfTurn() {
        addToBot(new RemoveSpecificPowerAction(owner, owner, this));
    }
}

File: src/main/kotlin/net/bindernews/grackle/power/TempThornsPower.kt
package net.bindernews.grackle.power

import com.megacrit.cardcrawl.actions.AbstractGameAction.AttackEffect.SLASH_HORIZONTAL
import com.megacrit.cardcrawl.actions.common.DamageAction
import com.megacrit.cardcrawl.actions.common.RemoveSpecificPowerAction
import com.megacrit.cardcrawl.cards.DamageInfo
import com.megacrit.cardcrawl.cards.DamageInfo.DamageType
import com.megacrit.cardcrawl.core.AbstractCreature
import net.bindernews.grackle.GrackleMod

class TempThornsPower(owner: AbstractCreature, amount: Int) : BasePower(POWER_ID) {
    init {
        setOwnerAmount(owner, amount)
        loadRegion("thorns")
        isTurnBased = true
        updateDescription()
    }

    override fun onAttacked(info: DamageInfo, damageAmount: Int): Int {
        val dt = info.type
        val own = info.owner
        if (dt != DamageType.THORNS && dt != DamageType.HP_LOSS && own != null && own !== owner) {
            flash()
            addToTop(DamageAction(own, DamageInfo(owner, amount, DamageType.THORNS), SLASH_HORIZONTAL, true))
        }
        return damageAmount
    }

    override fun updateDescription() {
        description = formatDesc(0, amount)
    }

    override fun atStartOfTurn() {
        addToBot(RemoveSpecificPowerAction(owner, owner, this))
    }

    companion object {
        @JvmField val POWER_ID = GrackleMod.makeId(TempThornsPower::class.java)
    }
}